pkg/registry/blobs.go

```diff
@@ -44,10 +48,90 @@ func isBlob(req *http.Request) bool {
 		elem[len(elem)-2] == "uploads")
 }
 
+// blobHandler represents a blob storage backend.
+//
+// For all methods, repo is a string that can be passed to
+// pkg/name.NewRepository to validate the repository.
+type blobHandler interface {
+	// Stat returns the size of the blob, or errNotFound if the blob wasn't found.
+	Stat(ctx context.Context, repo string, h v1.Hash) (int64, error)
+
+	// Get gets the blob contents, or errNotFound if the blob wasn't found.
+	Get(ctx context.Context, repo string, h v1.Hash) (io.ReadCloser, error)
+
+	// Put puts the blob contents.
+	//
+	// The contents will be verified against the expected size and digest
+	// as the contents are read, and an error will be returned if these
+	// don't match. Implementations should return that error, or a wrapper
+	// around that error, to return the correct error when these don't match.
+	Put(ctx context.Context, repo string, h v1.Hash, rc io.ReadCloser) error
+}
+
+// redirectError represents a signal that the blob handler doesn't have the blob
+// contents, but that those contents are at another location which registry
+// clients should redirect to.
+type redirectError struct {
```

Comment:
I kinda regret letting Scott use errors for non-error cases like this in `genreconciler`, but that ship sailed there.  Perhaps we can have the methods return a `Result` type or something, which can be an error, but don't have to be an error.

Reply:
I think I can justify this as a valid error scenario: "oops, I don't have it, but I know who does"

I could collapse this with `errNotFound`, so the same error type expresses "oops not found[, but I know who does]" if `Location` is set.

Reply:
4xx and 5xx are clear error cases, but 3xx feels like a reach.

I'm not gonna die on this hill, it just feels like a funny API.

Reply:
I don't disagree, it's weird. Luckily everything's still internal so we can change it to something we like better if we find it.